<a href="https://colab.research.google.com/github/ValeriaYashan/Machine-learning/blob/main/Riesgo_de_pobreza.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Este es un modelo de clasificación binaria que utiliza regresión logística para predecir el nivel de pobreza de una persona en función de sus características socioeconómicas.

Regresión Logística: Es un algoritmo de aprendizaje supervisado utilizado para predecir la probabilidad de que una observación pertenezca a una clase particular. En este caso, la regresión logística se utiliza para predecir la probabilidad de que una persona se encuentre en una categoría específica de pobreza (pobreza extrema, pobreza moderada o no pobreza) en función de sus ingresos medios.

Clasificación Binaria: El problema de clasificación se simplifica en este caso a una clasificación binaria entre "pobreza" y "no pobreza". La regresión logística calcula la probabilidad de pertenecer a la clase "pobreza" y, utilizando un umbral, asigna una etiqueta de clase a cada instancia.

Umbrales de Pobreza: Se definen umbrales de ingresos para determinar las categorías de pobreza. En este caso, se han definido dos umbrales: uno para la "pobreza extrema" y otro para la "pobreza moderada".

Evaluación del Modelo: Se evalúa el rendimiento del modelo utilizando métricas como la exactitud (accuracy) y el reporte de clasificación. La exactitud indica la proporción de predicciones correctas sobre el total de predicciones realizadas en el conjunto de prueba, mientras que el reporte de clasificación proporciona métricas adicionales como precision, recall y f1-score para cada clase.

In [69]:
import csv

# Datos completos
datos_completos = [
    {"Año": 2000, "Ingresos_medios": 25000, "Ingresos_mediana": 20000, "Ingresos_per_capita": 10000},
    {"Año": 2001, "Ingresos_medios": 26000, "Ingresos_mediana": 21000, "Ingresos_per_capita": 10500},
    {"Año": 2002, "Ingresos_medios": 27000, "Ingresos_mediana": 22000, "Ingresos_per_capita": 11000},
    {"Año": 2003, "Ingresos_medios": 28000, "Ingresos_mediana": 23000, "Ingresos_per_capita": 11500},
    {"Año": 2004, "Ingresos_medios": 29000, "Ingresos_mediana": 24000, "Ingresos_per_capita": 12000},
    {"Año": 2005, "Ingresos_medios": 30000, "Ingresos_mediana": 25000, "Ingresos_per_capita": 12500},
    {"Año": 2006, "Ingresos_medios": 31000, "Ingresos_mediana": 26000, "Ingresos_per_capita": 13000},
    {"Año": 2007, "Ingresos_medios": 32000, "Ingresos_mediana": 27000, "Ingresos_per_capita": 13500},
    {"Año": 2008, "Ingresos_medios": 33000, "Ingresos_mediana": 28000, "Ingresos_per_capita": 14000},
    {"Año": 2009, "Ingresos_medios": 34000, "Ingresos_mediana": 29000, "Ingresos_per_capita": 14500},
    {"Año": 2010, "Ingresos_medios": 35000, "Ingresos_mediana": 30000, "Ingresos_per_capita": 15000}
]

# Nombre del archivo CSV
archivo_csv = 'datos_ingresos.csv'

# Lista de nombres de las columnas
columnas = ['Año', 'Ingresos_medios', 'Ingresos_mediana', 'Ingresos_per_capita']

# Escribir los datos en el archivo CSV
with open(archivo_csv, 'w', newline='') as archivo:
    writer = csv.DictWriter(archivo, fieldnames=columnas)

    # Escribir los nombres de las columnas
    writer.writeheader()

    # Escribir los datos
    writer.writerows(datos_completos)

print("El archivo CSV 'datos_ingresos.csv' se ha creado exitosamente.")

El archivo CSV 'datos_ingresos.csv' se ha creado exitosamente.


In [70]:
import csv

# Nombre del archivo CSV
archivo_csv = 'datos.csv'

# Lista de nombres de las columnas
columnas = ['Año','Edad', 'Género', 'Nivel_educativo', 'Ingresos_familiares', 'Acceso_agua_potable', 'Acceso_electricidad']

# Datos a escribir en el archivo CSV
datos = [
    {"Año": 2000,'Edad': 25, 'Género': 'Masculino', 'Nivel_educativo': 'Universidad', 'Ingresos_familiares': 15000, 'Acceso_agua_potable': 'Sí', 'Acceso_electricidad': 'Sí'},
    {"Año": 2005,'Edad': 35, 'Género': 'Femenino', 'Nivel_educativo': 'Secundaria', 'Ingresos_familiares': 8000, 'Acceso_agua_potable': 'No', 'Acceso_electricidad': 'Sí'}
]

# Escribir los datos en el archivo CSV
with open(archivo_csv, 'w', newline='') as archivo:
    writer = csv.DictWriter(archivo, fieldnames=columnas)

    # Escribir los nombres de las columnas
    writer.writeheader()

    # Escribir los datos
    writer.writerows(datos)

print("El archivo CSV ha sido creado exitosamente.")

El archivo CSV ha sido creado exitosamente.


In [71]:
import csv

# Nombre del archivo CSV
archivo_csv = 'datos_rubro.csv'

# Lista de nombres de las columnas
columnas = ['Año','Rubro', 'Salario_minimo']

# Datos a escribir en el archivo CSV
datos = [
    {'Año':2005,'Rubro': 'Comercio', 'Salario_minimo': 40000},
    {'Año':2005,'Rubro': 'Manufactura', 'Salario_minimo': 38000},
    {'Año':2005,'Rubro': 'Construcción', 'Salario_minimo': 42000},
    {'Año':2005,'Rubro': 'Servicios', 'Salario_minimo': 37000},
    {'Año':2005,'Rubro': 'Agricultura', 'Salario_minimo': 35000}
]

# Escribir los datos en el archivo CSV
with open(archivo_csv, 'w', newline='') as archivo:
    writer = csv.DictWriter(archivo, fieldnames=columnas)

    # Escribir los nombres de las columnas
    writer.writeheader()

    # Escribir los datos
    writer.writerows(datos)

print("El archivo CSV ha sido creado exitosamente.")

El archivo CSV ha sido creado exitosamente.


SUBO LOS ARCHIVOS QUE NECESITO

In [40]:
from google.colab import files

# Subir los archivos desde tu computadora
uploaded = files.upload()

# Verificar que los archivos se hayan cargado correctamente
for filename in uploaded.keys():
    print('El archivo "{}" se ha cargado correctamente.'.format(filename))


Saving datos_ingresos.csv to datos_ingresos (1).csv
El archivo "datos_ingresos (1).csv" se ha cargado correctamente.


In [72]:
import pandas as pd

# Cargar los datos
datos_ingresos = pd.read_csv("datos_ingresos.csv")
datos_datos = pd.read_csv("datos.csv")
datos_rubro = pd.read_csv("datos_rubro.csv")

# Mostrar las primeras líneas del DataFrame
print("Primeras líneas del DataFrame de ingresos:")
print(datos_ingresos.head())
print()

print("Primeras líneas del DataFrame de datos:")
print(datos_datos.head())
print()

print("Primeras líneas del DataFrame de rubros:")
print(datos_rubro.head())



Primeras líneas del DataFrame de ingresos:
    Año  Ingresos_medios  Ingresos_mediana  Ingresos_per_capita
0  2000            25000             20000                10000
1  2001            26000             21000                10500
2  2002            27000             22000                11000
3  2003            28000             23000                11500
4  2004            29000             24000                12000

Primeras líneas del DataFrame de datos:
    Año  Edad     Género Nivel_educativo  Ingresos_familiares  \
0  2000    25  Masculino     Universidad                15000   
1  2005    35   Femenino      Secundaria                 8000   

  Acceso_agua_potable Acceso_electricidad  
0                  Sí                  Sí  
1                  No                  Sí  

Primeras líneas del DataFrame de rubros:
    Año         Rubro  Salario_minimo
0  2005      Comercio           40000
1  2005   Manufactura           38000
2  2005  Construcción           42000
3  2005   

MODELO

In [77]:
# Convertir variables categóricas en variables numéricas utilizando one-hot encoding
data_encoded = pd.get_dummies(data)

# Dividir los datos en características (X) y etiquetas (y) nuevamente
X_encoded = data_encoded.drop('Año', axis=1)
y_encoded = data_encoded['Año']

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train_encoded, X_test_encoded, y_train_encoded, y_test_encoded = train_test_split(X_encoded, y_encoded, test_size=0.2, random_state=42)

# Inicializar y entrenar el modelo de regresión lineal con los datos codificados
model_encoded = LinearRegression()
model_encoded.fit(X_train_encoded, y_train_encoded)

# Predecir en el conjunto de prueba
y_pred_encoded = model_encoded.predict(X_test_encoded)

# Evaluar el rendimiento del modelo
mse_encoded = mean_squared_error(y_test_encoded, y_pred_encoded)
r2_encoded = r2_score(y_test_encoded, y_pred_encoded)
print('Error cuadrático medio:', mse_encoded)
print('Coeficiente de determinación:', r2_encoded)

Error cuadrático medio: 0.0
Coeficiente de determinación: nan


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:918: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)



Este fragmento de código realiza los siguientes pasos:

One-Hot Encoding: Utiliza la función pd.get_dummies() para convertir las variables categóricas en variables numéricas utilizando one-hot encoding. Esto crea nuevas columnas binarias para cada valor único en las variables categóricas.

División de los datos codificados: Divide los datos transformados en características (X) y etiquetas (y) nuevamente.

División en conjuntos de entrenamiento y prueba: Divide los datos codificados en conjuntos de entrenamiento y prueba utilizando la función train_test_split().

Inicialización y entrenamiento del modelo de regresión lineal con los datos codificados: Inicializa un modelo de regresión lineal y lo entrena utilizando los datos codificados de entrenamiento.

Predicción en el conjunto de prueba: Utiliza el modelo entrenado para hacer predicciones sobre el conjunto de prueba codificado.

Evaluación del rendimiento del modelo: Calcula el error cuadrático medio (MSE) y el coeficiente de determinación (R^2) para evaluar el rendimiento del modelo en el conjunto de prueba codificado.

En resumen, este código aplica one-hot encoding a los datos categóricos, entrena un modelo de regresión lineal con los datos transformados y evalúa su rendimiento utilizando métricas de regresión.

Predecir pobreza

In [76]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Cargar los datos
datos = pd.read_csv("datos_ingresos.csv")

# Definir los umbrales de pobreza
umbral_extrema = 30000
umbral_moderada = 50000

# Etiquetar las personas según su nivel de pobreza
datos['Pobreza'] = pd.cut(datos['Ingresos_medios'], bins=[-float('inf'), umbral_extrema, umbral_moderada, float('inf')], labels=['pobreza_extrema', 'pobreza_moderada', 'no_pobreza'])

# Dividir los datos en características (X) y etiquetas (y)
X = datos.drop(['Año', 'Pobreza'], axis=1)
y = datos['Pobreza']

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Inicializar y entrenar el modelo de regresión logística
modelo = LogisticRegression()
modelo.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred = modelo.predict(X_test)

# Evaluar el rendimiento del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Exactitud del modelo:", accuracy)

# Mostrar el reporte de clasificación
print("\nReporte de Clasificación:")
print(classification_report(y_test, y_pred))

Exactitud del modelo: 0.6666666666666666

Reporte de Clasificación:
                  precision    recall  f1-score   support

 pobreza_extrema       1.00      0.50      0.67         2
pobreza_moderada       0.50      1.00      0.67         1

        accuracy                           0.67         3
       macro avg       0.75      0.75      0.67         3
    weighted avg       0.83      0.67      0.67         3



Estos resultados indican el rendimiento del modelo de clasificación que he entrenado.

La "Exactitud del modelo" indica la proporción de predicciones correctas sobre el total de predicciones realizadas en el conjunto de prueba. En este caso, el modelo tiene una exactitud del 66.67%, lo que significa que clasificó correctamente el 66.67% de las instancias del conjunto de prueba.

El "Reporte de Clasificación" proporciona una descripción más detallada del rendimiento del modelo para cada clase. Se divide en cuatro métricas principales:

Precision: La proporción de instancias clasificadas como positivas (verdaderos positivos) entre todas las instancias clasificadas como positivas por el modelo.
Recall: La proporción de instancias clasificadas como positivas (verdaderos positivos) entre todas las instancias que realmente son positivas en los datos.
F1-score: Una medida ponderada de precision y recall, que equilibra ambas métricas. Es útil cuando las clases están desbalanceadas.
Support: El número de instancias de cada clase en el conjunto de prueba.
El modelo muestra buenos resultados en cuanto a precision y recall para ambas clases ("pobreza_extrema" y "pobreza_moderada"). Sin embargo, se debe tener en cuenta que estos resultados se basan en un conjunto de prueba muy pequeño (3 instancias), por lo que podrían no ser representativos del rendimiento real del modelo en datos más grandes y diversos.